src/web-extensions.c

```diff
@@ -0,0 +1,696 @@
+}
+
+static GStrv
+get_manifest_search_path (void)
+{
+  const char *hosts_path_str;
+  g_autoptr(GPtrArray) search_path = NULL;
+
+  hosts_path_str = g_getenv ("XDG_DESKTOP_PORTAL_WEB_EXTENSIONS_PATH");
+  if (hosts_path_str != NULL)
+    return g_strsplit (hosts_path_str, ":", -1);
+
+  /* By default, use the native messaging search paths of Firefox,
+   * Chrome, and Chromium, as documented here:
+   *
+   * https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#manifest_location
+   * https://developer.chrome.com/docs/apps/nativeMessaging/#native-messaging-host-location
+   */
+
+  search_path = g_ptr_array_new_with_free_func (g_free);
+  /* Add per-user directories */
+  g_ptr_array_add (search_path, g_build_filename (g_get_home_dir (), ".mozilla", "native-messaging-hosts", NULL));
+  g_ptr_array_add (search_path, g_build_filename (g_get_user_config_dir (), "google-chrome", "NativeMessagingHosts", NULL));
+  g_ptr_array_add (search_path, g_build_filename (g_get_user_config_dir (), "chromium", "NativeMessagingHosts", NULL));
+
+  /* Add system wide directories */
+  g_ptr_array_add (search_path, g_strdup ("/usr/lib/mozilla/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup ("/usr/lib64/mozilla/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup ("/etc/opt/chrome/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup ("/etc/chromium/native-messaging-hosts"));
+
+  /* And the same for xdg-desktop-portal's configured prefix */
+  g_ptr_array_add (search_path, g_strdup (LIBDIR "mozilla/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup (SYSCONFDIR "opt/chrome/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup (SYSCONFDIR "chromium/native-messaging-hosts"));
+
+  g_ptr_array_add (search_path, NULL);
+  return (GStrv)g_ptr_array_free (g_steal_pointer (&search_path), FALSE);
+}
+
+static char *
+find_server (const char *server_name,
+             const char *extension_or_origin,
+             char **server_description,
+             char **json_manifest,
+             GError **error)
+{
+  g_auto(GStrv) search_path;
+  g_autoptr(JsonParser) parser = NULL;
+  g_autofree char *metadata_basename = NULL;
+  int i;
+
+  /* Check that the we have a valid native messaging host name */
+  if (!is_valid_name (server_name))
+    {
+      g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT, "invalid native messaging server name");
+      return NULL;
+    }
+
+  search_path = get_manifest_search_path ();
+  parser = json_parser_new ();
+  metadata_basename = g_strconcat (server_name, ".json", NULL);
+
+  for (i = 0; search_path[i] != NULL; i++)
+    {
+      g_autofree char *metadata_filename = NULL;
+      JsonObject *metadata_root;
+
+      metadata_filename = g_build_filename (search_path[i], metadata_basename, NULL);
+      if (!g_file_test (metadata_filename, G_FILE_TEST_EXISTS))
+        continue;
```

Comment:
TOCTOU potential here. Maybe this particular check should be skipped, and "replaced" by more sophisticated error handling in the `json_parser_load_from_file` call below?